diracx-testing/src/diracx/testing/client_generation.py

```diff
@@ -0,0 +1,140 @@
+            }
+    raise NotImplementedError("__all__ not found")
+
+
+def fixup_models_init(generated_dir, extension_name):
+    """Workaround for https://github.com/python/mypy/issues/15300."""
+    models_init_path = generated_dir / "models" / "__init__.py"
+    # Enums cannot be extended, so we don't need to patch them
+    object_names = {
+        name
+        for name, module in extract_static_all(models_init_path).items()
+        if module != "_enums"
+    }
+
+    patch_module = "diracx.client._generated.models._patch"
+    spec = importlib.util.find_spec(patch_module)
+    if spec is None or spec.origin is None:
+        raise ImportError(f"Cannot locate {patch_module} package")
+    missing = set(extract_static_all(Path(spec.origin))) - set(object_names)
+    missing_formatted = "\n".join(f'            "{name}",' for name in missing)
+
+    with models_init_path.open("a") as fh:
+        fh.write(
+            "if TYPE_CHECKING:\n"
+            "    __all__.extend(\n"
+            "        [\n"
+            f"{missing_formatted}"
+            "        ]\n"
+            "    )\n"
+        )
+
+
+def regenerate_client(openapi_spec: Path, client_root: Path):
+    """Regenerate the AutoREST client and run pre-commit checks on it.
+
+    This test is skipped by default, and can be enabled by passing
+    --regenerate-client to pytest. It is intended to be run manually
+    when the API changes.
+
+    The reason this is a test is that it is the only way to get access to the
+    test_client fixture, which is required to get the OpenAPI spec.
+
+    WARNING: This test will modify the source code of the client!
+    """
+    assert client_root.is_dir()
+    assert client_root.name == "client"
+    assert (client_root / "_generated").is_dir()
+    extension_name = client_root.parent.name
+
+    repo_root = client_root.parents[3]
+    if extension_name == "gubbins":
+        # Gubbins is special because it has a different structure due to being
+        # in a subdirectory of diracx
+        repo_root = repo_root.parents[1]
+    assert (repo_root / ".git").is_dir()
+    repo = git.Repo(repo_root)
+    generated_dir = client_root / "_generated"
+    if repo.is_dirty(path=generated_dir):
+        raise AssertionError(
+            "Client is currently in a modified state, skipping regeneration"
+        )
+
+    cmd = [
+        "autorest",
+        "--python",
+        f"--input-file={openapi_spec}",
+        "--models-mode=msrest",
+        "--namespace=_generated",
+        f"--output-folder={client_root}",
+    ]
+
+    # This is required to be able to work offline
+    # TODO: if offline, find the version already installed
+    # and use it
+    # cmd += [f"--use=@autorest/python@{AUTOREST_VERSION}"]
+
+    # ruff: disable=S603
+    subprocess.run(cmd, check=True)
+
+    if extension_name != "diracx":
+        # For now we don't support extending the models in extensions. To make
```

Comment:
Why not ? We said we would just do a super set of all of them

Reply:
It's included in https://github.com/DIRACGrid/diracx/issues/442 (though I'm actually not convinced it will ever be used...)